### ISSUE TITLE
Is C# 9 supported?

### ISSUE THREAD

**Issue by WhiteBlackGoose** [doxygen/doxygen#8428]
Hello. I was trying to build docs for my project which relies on C# 9 syntax features. But Doxygen skipped all my `record`s with error
```
error: NamespaceDefImpl::insertMembers(): member 'MyRecord' with class scope '' inserted in namespace scope 'MyNamespace'!
```
So the question is whether C# 9 is supported, and if it is, how do I enable it?

**Comment by albert-github** [comment]
C# 9 is quite new and as far as I can see there is no official standard for it only a specification proposal from Microsoft (seen on https://en.wikipedia.org/wiki/C_Sharp_(programming_language) ).
Most likely it is not supported.

- Do you have an authoritative source to a standard?
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.  
- Please also specify the full doxygen version used (`doxygen -v`).

**Comment by WhiteBlackGoose** [comment]
> Do you have an authoritative source to a standard?

The [proposals](https://github.com/dotnet/csharplang/tree/main/proposals/csharp-9.0) pretty much describe all the features with antlr syntax and specifications.

> Can you please attach a, small, self contained example

[DoxyGenCSharp9Bug.zip](https://github.com/doxygen/doxygen/files/6136357/DoxyGenCSharp9Bug.zip)

> Please also specify the full doxygen version used

1.9.1 (ef9b20ac7f8a8621fcfc299f8bd0b80422390f4b)

Notes about the example:
1. It treats a record with parameters with no methods as a function.
2. It treats a record with parameters and with methods as a function, ignoring the methods
3. It treats a record with no parameters as a variable
4. It treats a partial record with no parameters with an error, ignoring in the resulting documentation

**Comment by albert-github** [comment]
Thanks for the link.
The disadvantage, unfortunately, is that it is, to what I understand of the structure, just proposed changes compared to the previous version  and not one document so one has to look through previous versions as well to get a grip on it.
In the `spec` directory there is a draft version of the 6.0 standard and this says:
> This is an unofficial draft, posted here for convenience. When we submit a C# 6.0 spec proposal to ECMA, it will be shared here.

looks like not a lot of activity going on in this directory, except for some cosmetic changes.
- do you know when the 6.0 draft will be send to ECMA for standardization?
- do you know how far this draft spec is behind the proposals for the different versions C#9 / C# 8 ...?

**Comment by WhiteBlackGoose** [comment]
> do you know when the 6.0 draft will be send to ECMA for standardization?

No, I don't know when. Is it required for the syntax to be implemented for Doxygen?

> do you know how far this draft spec is behind the proposals for the different versions C#9 / C# 8 ...?

I assume that what's been implemented since then in terms of what is usually documented are `records` only. At least, after going through the versions I couldn't find anything that Doxy fails to generate, with except for `records`.

**Comment by albert-github** [comment]
> > do you know when the 6.0 draft will be send to ECMA for standardization?

> No, I don't know when. Is it required for the syntax to be implemented for Doxygen?

Strictly not when it is a stable definition, but with not official standardized new features one never knows (and even with official standardized features it can happen that in time things change, as I recollect this happened with a C++ feature).

**Comment by WhiteBlackGoose** [comment]
It is surely not going to change. All syntactical features will be reflected in newer versions, not in older ones. They can fix bugs in the runtime (CLR), but it's not related to syntax.

If there's no plans to implement `records`'s documentation generation, it's very unfortunate, and I will have to look for another doc-generator :(.

**Comment by albert-github** [comment]
> If there's no plans to implement records's documentation generation, it's very unfortunate, and I will have to look for another doc-generator :(.

It is nowhere written it is not going to be implemented its is probably just that someone will find time to implement it, feel free to create a proposed pull request for doxygen.

**Comment by albert-github** [comment]
Looks like after numerous years a new version of the C# Language Definition (Standard ECMA-334 6th Edition i.e. https://www.ecma-international.org/wp-content/uploads/ECMA-334_6th_edition_june_2022.pdf) has been released.
